### ISSUE TITLE
Function parameter type info gets lost when called with .always_inline

### ISSUE THREAD

**Issue by psnszsn** [ziglang/zig#18113]
### Zig Version

0.12.0-dev.1718+3acb0e30a

### Steps to Reproduce and Observed Behavior

```zig
fn asdf(ptr: *u32) void {
    std.debug.print("value: {}", .{ptr.*});
}
test {
    const fn_ptr: *const fn (*anyopaque) void = @ptrCast(&asdf);
    var a: u32 = 77;
    @call(.always_inline, fn_ptr, .{@as(*anyopaque, @ptrCast(&a))});
}
```

Error message: 
```
src/main.zig:153:39: error: cannot load opaque type 'anyopaque'
    std.debug.print("value: {}", .{ptr.*});
                                   ~~~^~
src/main.zig:158:5: note: called from here
    @call(.always_inline, fn_ptr, .{@as(*anyopaque, @ptrCast(&a))});
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

If I call the function with `.auto` the function is called correctly and no error message is shown.

### Expected Behavior

I would expect the function parameters to be automatically coerced even when the function is inlined.

**Comment by xdBronch** [comment]
this seems like technically correct behavior since youre not calling `asdf`, youre calling a function pointer thats been casted to take `*anyopaque`. arguably you shouldnt be allowed to inline a function pointer which is the cause of this.

**Comment by psnszsn** [comment]
> arguably you shouldn't be allowed to inline a function pointer which is the cause of this.

Calling the function pointer with `.always_inline` requires it to be comptime known. 

Requiring a comptime-known function body type instead would make it impossible to cast the function type, I think. 
At least I have not been able to cast between function body types.

**Comment by xdBronch** [comment]
i know and thats in part my point, why would you do this? how would this work? `ptr` just is not a `*u32` anymore. if anyone knows better than me please correct me but if you want this to work i think you need a language proposal, not a bug fix

**Comment by Rexicon226** [comment]
> i know and thats in part my point, why would you do this? how would this work? `ptr` just is not a `*u32` anymore. if anyone knows better than me please correct me but if you want this to work i think you need a language proposal, not a bug fix

absolutely... the second you casted to a function prototype with a type-erased argument, that's it. the type is gone, and the `@call` correctly errors in that it cannot "re-type" something that's erased. 

the cast happens first, then the call :smile:

**Comment by psnszsn** [comment]
There are plenty of reasons for type-erasing the function parameter. One example would be handling user defined callbacks.

This is something that works in Zig when not inlining the function. Just replace `.always_inline` with `.auto` in my example. The pointer does not get "re-typed" since there are no types at runtime, but the function is called with a valid argument and it prints "77" as expected.

I would expect the same thing to happen when the function is inlined (assuming the function pointer is comptime known ofc). I guess the inlining has to be done later in the compilation pipeline, after type checking.

**Comment by xdBronch** [comment]
if youre taking in a `*u32` and you expect it to behave like a `*u32` surely it has to be a `*u32` thats passed in? if im understanding the use case you mentioned, thats usually done more like this where the type is opaque to the library at every stage
```zig
fn userCallback(data: *anyopaque) void {
    const user_data: UserType = @ptrCast(ud);
}
```
ive never seen it done like what youre showing and it seems extremely odd to do so, i really wouldnt expect it to work

**Comment by Rexicon226** [comment]
to add to @xdBronch, if you **really** want to use `anyopaque`, you could add a `ptrCast` to the function....

this is very silly and just goes around the entire purpose of the function prototype?

second: after some discussion on the discord there is #18117, which will simply prevent this from being done. inlining a function pointer is very strange behavior that should not be done.
